File: bridge1/src/uhf-sidecar.js
'use strict';

/**
 * Sidecar driver: same surface as uhf.js (the DLL driver), but every call is
 * an HTTP request to a sidecar process that owns the reader. Two servers speak
 * the contract:
 *   - bridge/sidecar/UhfSidecar.java — pure-Java SDK, for Linux / Raspberry Pi
 *     where the Windows DLL cannot load (network readers only);
 *   - bridge/src/sidecar-server.js  — uhf.js over HTTP, for a USB/COM desktop
 *     reader plugged into a DIFFERENT Windows PC than the bridge.
 *
 * Differences from the DLL driver:
 *   - all functions are async (callers already run them inside awaited locks)
 *   - capabilities.hw = false: HW trigger mode (work mode 2 + UDP push) is
 *     not implemented in the sidecar yet — IR (bridge) and manual modes only.
 *   - drainTags() replaces the pollTag() loop: tags accumulate in the
 *     sidecar's queue and are fetched in batches.
 */

const { spawn } = require('child_process');
const path = require('path');

const SIDECAR_URL = process.env.UHF_SIDECAR_URL || 'http://127.0.0.1:3010';
const SIDECAR_DIR = path.join(__dirname, '..', 'sidecar');
const JAR = 'ReaderAPI20240822.jar';

// A remote sidecar (reader plugged into ANOTHER machine, running
// sidecar-server.js) cannot be spawned from here — only a local one can.
const SIDECAR_IS_LOCAL = ['127.0.0.1', 'localhost', '::1'].includes(new URL(SIDECAR_URL).hostname);

const capabilities = { hw: false, ioStatusDebug: false };

let child = null;

async function call(method, pathname, params = {}) {
  const url = new URL(pathname, SIDECAR_URL);
  for (const [k, v] of Object.entries(params)) if (v != null) url.searchParams.set(k, String(v));
  const res = await fetch(url, { method });
  return res.json();
}

/**
 * Is the thing on the port actually driving a reader?
 *
 * /status alone is NOT enough, and trusting it cost a shift of downtime: a
 * sidecar left over from a previous bridge kept answering {ok:true,
 * connected:true} long after it had lost the reader, so every hardware call
 * failed while the bridge cheerfully reported "reading..." and the antenna
 * lights were dark. A hardware call is the only honest probe.
 */
async function sidecarIsHealthy() {
  try {
    const status = await call('GET', '/status');
    if (!status?.ok) return false;
    // A sidecar that has not been asked to connect yet is perfectly healthy —
    // it simply has no reader handle, and its hardware calls fail for that
    // reason alone. This is the normal state at boot, BEFORE POST /connect.
    if (!status.connected) return true;
    // It claims a live reader, so make it prove it. The stale-orphan signature
    // is exactly this contradiction: connected:true while every hardware call
    // returns ok:false. That is what kept the antenna lights dark while the
    // bridge logged "reading..." for hours.
    const probe = await call('GET', '/antennas');
    return probe?.ok === true;
  } catch (_) {
    return false;
  }
}

/** Spawn the Java sidecar if nothing healthy is listening yet. Idempotent. */
async function ensureSidecar() {
  if (await sidecarIsHealthy()) return; // already up AND working

  // Something may still be holding the port without being usable — an orphan
  // from a bridge that exited without taking its child with it. We cannot
  // adopt it and we cannot bind over it, so say precisely what to do rather
  // than failing later with a symptom ("no tags") far from the cause.
  let occupied = false;
  try {
    const status = await call('GET', '/status');
    // Only a CONNECTED-but-broken sidecar is unusable. One that answers and
    // admits it holds no reader is just waiting for /connect.
    occupied = Boolean(status?.connected);
  } catch (_) {
    /* genuinely nothing listening — the normal path */
  }
  if (occupied) {
    throw new Error(
      `a sidecar is listening on ${SIDECAR_URL} but its reader calls fail — it is a stale process from an earlier bridge. ` +
        `Stop it and start the bridge again:  pkill -f UhfSidecar`
    );
  }
  if (!SIDECAR_IS_LOCAL) {
    throw new Error(
      `sidecar at ${SIDECAR_URL} is not answering — start sidecar-server.js on the reader PC (and check its firewall allows this port)`
    );
  }
  const sep = process.platform === 'win32' ? ';' : ':';
  child = spawn('java', ['-cp', `${JAR}${sep}.`, 'UhfSidecar', String(new URL(SIDECAR_URL).port || 3010)], {
    cwd: SIDECAR_DIR,
    stdio: ['ignore', 'pipe', 'pipe'],
  });
  child.stdout.on('data', (d) => console.log(`[sidecar] ${String(d).trim()}`));
  child.stderr.on('data', (d) => console.error(`[sidecar] ${String(d).trim()}`));
  child.on('exit', (code) => {
    console.error(`[sidecar] exited with code ${code}`);
    child = null;
  });
  // The sidecar is supposed to live and die with the bridge. It did not: a
  // killed bridge left it running, systemd adopted it, and the next bridge
  // attached to a corpse. Tie the lifetimes together explicitly.
  if (!ensureSidecar._exitHooked) {
    ensureSidecar._exitHooked = true;
    const stopChild = () => {
      if (child && !child.killed) {
        try {
          child.kill('SIGTERM');
        } catch (_) {
          /* already gone */
        }
      }
    };
    process.on('exit', stopChild);
    for (const sig of ['SIGINT', 'SIGTERM', 'SIGHUP']) {
      process.on(sig, () => {
        stopChild();
        process.exit(0);
      });
    }
  }
  // wait for it to come up
  for (let i = 0; i < 20; i++) {
    await new Promise((r) => setTimeout(r, 250));
    try {
      await call('GET', '/status');
      return;
    } catch (_) {
      /* retry */
    }
  }
  throw new Error('sidecar did not start (is java on PATH? jar compiled?)');
}

function load() {
  return ensureSidecar();
}

async function connect(ip, port) {
  await ensureSidecar();
  const r = await call('POST', '/connect', { ip, port });
  return r.ok ? 0 : 2; // mimic SDK rc: 0 ok, 2 connect failure
}

/**
 * Open a USB/COM desktop reader plugged into the SIDECAR's machine. Served by
 * sidecar-server.js (Windows, DLL); the Java/Linux sidecar has no such route
 * and answers non-ok, which surfaces as rc 2 rather than a thrown mystery.
 */
async function connectUsb() {
  await ensureSidecar();
  const r = await call('POST', '/connect-usb');
  return r.ok ? 0 : 2;
}

async function disconnect() {
  try {
    await call('POST', '/disconnect');
  } catch (_) {
    /* sidecar gone = disconnected enough */
  }
}

async function startInventory() {
  const r = await call('POST', '/inventory/start');
  return r.ok ? 0 : -1;
}

async function stopInventory() {
  const r = await call('POST', '/inventory/stop');
  return r.ok ? 0 : -1;
}

/** Fetch up to max tags. Returns uhf.js-shaped tag objects. */
async function drainTags(max = 100) {
  const r = await call('GET', '/tags', { max });
  if (!r.ok || !Array.isArray(r.tags)) return [];
  return r.tags.map((t) => ({
    epc: t.epc ? String(t.epc).toUpperCase() : null,
    pc: t.pc || null,
    tid: t.tid || null,
    user: t.user || null,
    antenna: t.ant != null ? parseInt(t.ant, 10) : null,
    rssi: t.rssi != null ? parseFloat(t.rssi) : null,
    raw: null, // sidecar delivers parsed tags; no raw frame
  }));
}

/**
 * GPI read. Sidecar reports state 0/1 per input; on this hardware GPI1 idles
 * HIGH (beam clear) and drops LOW when broken -> broken = state 0.
 */
async function getGpi() {
  try {
    const r = await call('GET', '/gpi');
    if (!r.ok || !Array.isArray(r.gpi)) return { gpi1: null, gpi2: null, raw: '', rc: -1 };
    const byName = {};
    for (const g of r.gpi) byName[g.name] = g.state;
    const broken = (s) => (s == null ? null : s === 0);
    const raw = r.gpi.map((g) => String(g.state).padStart(2, '0')).join('');
    return { gpi1: broken(byName.GPI1), gpi2: broken(byName.GPI2), raw, rc: 0 };
  } catch (_) {
    return { gpi1: null, gpi2: null, raw: '', rc: -1 };
  }
}

async function getPower() {
  const r = await call('GET', '/power');
  if (!r.ok || !Array.isArray(r.power) || !r.power.length) return null;
  return r.power[0].dbm;
}

async function setPower(dBm, save = true) {
  const ants = (await getAntennas()) || [1];
  // `save` rides along for sidecar-server.js (session-only bench bumps must
  // not persist); the Java sidecar ignores unknown params, unchanged there.
  const r = await call('POST', '/power', { dbm: dBm, ants: ants.join(','), save: save ? 1 : 0 });
  return r.ok ? 0 : -1;
}

/**
 * Set power on ONE antenna port. The sidecar's /power already takes an `ants`
 * list — setPower above just happens to pass every enabled port at once — so
 * per-port control needs no new sidecar endpoint, only a narrower call.
 *
 * This matters at a gate because the ports are not equivalent: one antenna
 * covers the doorway, another reaches down the aisle, and running both at the
 * same dBm is what drags distant stock into the read zone while the carton
 * actually passing reads no better.
 */
async function setAntennaPower(port, dBm, save = true) {
  const r = await call('POST', '/power', { dbm: dBm, ants: String(port), save: save ? 1 : 0 });
  return r.ok ? 0 : -1;
}

/** Power per antenna. @returns {{[port:number]:{read:number,write:number}}|null} */
async function getAntennaPower() {
  const r = await call('GET', '/power');
  if (!r.ok || !Array.isArray(r.power)) return null;
  const out = {};
  for (const e of r.power) {
    // The sidecar has carried more than one key name for the port over time;
    // accept any of them rather than silently returning an empty map.
    const port = e?.ant ?? e?.antenna ?? e?.port;
    if (port == null || e?.dbm == null) continue;
    out[port] = { read: e.dbm, write: e.dbm };
  }
  return Object.keys(out).length ? out : null;
}

/**
 * Reader buzzer on/off. The UR4 chirps on every tag read by default, which at a
 * gate reading continuously is one long tone all shift.
 *
 * Returns 0 on success, -1 if the reader refused. Throws only if the sidecar
 * has no /beep route at all — an OLD sidecar build, which the caller reports as
 * "rebuild the sidecar" rather than as a reader fault.
 */
async function setBeep(on) {
  const r = await call('POST', '/beep', { on: on ? 1 : 0 });
  return r.ok ? 0 : -1;
}

/** @returns {boolean|null} null = the reader would not say. */
async function getBeep() {
  const r = await call('GET', '/beep');
  return r.ok ? Boolean(r.on) : null;
}

async function getAntennas() {
  const r = await call('GET', '/antennas');
  return r.ok ? r.enabled : null;
}

async function setAntennas(ports) {
  const r = await call('POST', '/antennas', { ports: ports.join(',') });
  return r.ok ? 0 : -1;
}

async function getAntennaLink() {
  return null; // not exposed by the Java SDK's network API
}

async function getWorkMode() {
  const r = await call('GET', '/workmode');
  return r.ok ? r.mode : null;
}

async function setWorkMode(mode) {
  const r = await call('POST', '/workmode', { mode });
  return r.ok ? 0 : -1;
}

async function getSoftwareVersion() {
  try {
    const r = await call('GET', '/version');
    return r.version || null;
  } catch (_) {
    return null;
  }
}

function setLogLevel() {}

// --- single-tag access (read / write / singulate) ---------------------------
// Same surface + shapes as uhf.js so server.js's /tag routes work unchanged on
// Linux. Filters are the same { bank, startBit, data:Buffer } objects; the
// sidecar receives them as fbank/fptr(bits)/fdata(hex) query params.

const BANK = { RESERVED: 0, EPC: 1, TID: 2, USER: 3 };
const EPC_DATA_PTR = 2;
const EPC_FILTER_BIT_OFFSET = 32; // CRC(16) + PC(16)

function hexToBuf(hex, label) {
  const clean = String(hex ?? '').replace(/[^0-9a-fA-F]/g, '');
  if (clean.length % 2 !== 0) throw new Error(`${label}: odd number of hex digits ("${hex}")`);
  return Buffer.from(clean, 'hex');
}

function filterByTid(tidHex) {
  const data = hexToBuf(tidHex, 'filterByTid');
  if (!data.length) throw new Error('filterByTid: empty TID');
  return { bank: BANK.TID, startBit: 0, data };
}

function filterByEpc(epcHex, startBit = EPC_FILTER_BIT_OFFSET) {
  const data = hexToBuf(epcHex, 'filterByEpc');
  if (!data.length) throw new Error('filterByEpc: empty EPC');
  return { bank: BANK.EPC, startBit, data };
}

function filterParams(filter) {
  if (!filter) return {};
  return {
    fbank: filter.bank,
    fptr: filter.startBit,
    fdata: filter.data.toString('hex').toUpperCase(),
  };
}

/** @returns {{rc:number, hex:(string|null), bytes:(Buffer|null)}} — mirrors uhf.js */
async function readBank({ bank, ptr = 0, words = 1, filter = null, accessPwd } = {}) {
  const r = await call('POST', '/tag/read', { bank, ptr, words, pwd: accessPwd, ...filterParams(filter) });
  if (!r.ok || !r.hex) return { rc: -1, hex: null, bytes: null };
  const hex = String(r.hex).toUpperCase();
  return { rc: 0, hex, bytes: Buffer.from(hex, 'hex') };
}

/** @returns {number} 0 on success — mirrors uhf.js */
async function writeBank({ bank, ptr = 0, dataHex, filter = null, accessPwd } = {}) {
  const clean = String(dataHex ?? '').replace(/[^0-9a-fA-F]/g, '').toUpperCase();
  if (!clean.length) throw new Error('writeBank: no data');
  if (clean.length % 4 !== 0) throw new Error(`writeBank: data must be whole 16-bit words (got ${clean.length / 2} bytes)`);
  const r = await call('POST', '/tag/write', { bank, ptr, data: clean, pwd: accessPwd, ...filterParams(filter) });
  return r.ok ? 0 : -1;
}

/** Singulate ONE tag. @returns {{pc:string, epc:string, raw:string}|null} */
async function inventorySingle() {
  const r = await call('GET', '/tag/single');
  if (!r.ok || !r.tag || !r.tag.epc) return null;
  const pc = String(r.tag.pc || '').toUpperCase();
  const epc = String(r.tag.epc).toUpperCase();
  return { pc, epc, raw: pc + epc };
}

/** EPC length in 16-bit words, as declared by a PC word (pure JS, same as uhf.js). */
function epcWordsFromPc(pcHex) {
  const buf = hexToBuf(pcHex, 'epcWordsFromPc');
  if (buf.length !== 2) return null;
  return buf.readUInt16BE(0) >> 11;
}

/** Liveness for _pollAlive: the reader answers a version query or it is gone. */
async function isReaderAlive() {
  return Boolean(await getSoftwareVersion());
}

const unsupported = (name) => () => {
  throw new Error(`${name} not supported by the sidecar driver (DLL driver only)`);
};

module.exports = {
  capabilities,
  load,
  connect,
  connectUsb,
  disconnect,
  isConnected: () => true, // controller tracks its own state
  isReaderAlive,
  startInventory,
  stopInventory,
  drainTags,
  getGpi,
  getPower,
  setBeep,
  getBeep,
  setAntennaPower,
  getAntennaPower,
  setPower,
  getAntennas,
  setAntennas,
  getAntennaLink,
  getWorkMode,
  setWorkMode,
  getSoftwareVersion,
  setLogLevel,
  // single-tag access (encode/read/write) — served by the Java sidecar
  BANK,
  EPC_DATA_PTR,
  filterByTid,
  filterByEpc,
  readBank,
  writeBank,
  inventorySingle,
  epcWordsFromPc,
  // debug/HW-mode surface not available via sidecar:
  readIOStatus: async () => null, // /debug/io still answers (gpi only) on Linux
  getIOControl: unsupported('getIOControl'),
  setGpiConfig: () => ({}),
  getGpiConfig: () => ({ source: 'sidecar' }),
  setWorkModePara: unsupported('setWorkModePara'),
  getWorkModePara: async () => null,
  setDestIp: unsupported('setDestIp'),
  getDestIp: async () => null,
  bindUdp: unsupported('bindUdp'),
  unbindUdp: () => {},
  parseUdpDatagram: () => null,
  getWorkTime: async () => null,
  setWorkTime: unsupported('setWorkTime'),
  setReaderIp: unsupported('setReaderIp'),
  getReaderIp: async () => null,
};

File: bridge1/sidecar/UhfSidecar.java
import com.rscja.deviceapi.RFIDWithUHFNetworkUR4;
import com.rscja.deviceapi.entity.AntennaNameEnum;
import com.rscja.deviceapi.entity.AntennaPowerEntity;
import com.rscja.deviceapi.entity.AntennaState;
import com.rscja.deviceapi.entity.GPIStateEntity;
import com.rscja.deviceapi.entity.UHFTAGInfo;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.Executors;

/**
 * UHF sidecar: exposes the Chainway Java SDK (pure-Java TCP path — works on
 * Windows AND Raspberry Pi/ARM) as a tiny local HTTP API for the Node bridge.
 *
 * Build:  javac -cp ReaderAPI20240822.jar UhfSidecar.java com/rscja/deviceapi/AntennaLink.java
 * Run:    java -cp "ReaderAPI20240822.jar;." UhfSidecar [port]     (Windows)
 *         java -cp "ReaderAPI20240822.jar:." UhfSidecar [port]     (Linux)
 *
 * All responses are JSON. Endpoints:
 *   POST /connect?ip=&port=      GET /status          POST /disconnect
 *   POST /inventory/start        POST /inventory/stop
 *   GET  /tags?max=200           (drains the callback queue)
 *   GET  /gpi                    GET /version
 *   GET  /power                  POST /power?dbm=&ants=1,2
 *   GET  /antennas               POST /antennas?ports=1,3
 *   GET  /antennas/link          (which ports have an antenna plugged in)
 *   GET  /workmode               POST /workmode?mode=0
 *   GET  /beep                   POST /beep?on=0        (0 = silence the reader)
 *   GET  /tag/single             (singulate ONE tag: pc + epc)
 *   POST /tag/read?bank=&ptr=&words=[&fbank=&fptr=&fdata=][&pwd=]
 *   POST /tag/write?bank=&ptr=&data=[&fbank=&fptr=&fdata=][&pwd=]
 *     Filter fields mirror the DLL driver: fptr is the mask offset in BITS,
 *     fdata is hex; the mask length is derived from fdata. No fdata = no
 *     filter (any tag in the field may answer — single-tag bench use only).
 */
public class UhfSidecar {
    static final RFIDWithUHFNetworkUR4 uhf = new RFIDWithUHFNetworkUR4();
    static final ConcurrentLinkedQueue<String> tagQueue = new ConcurrentLinkedQueue<>();
    static volatile boolean connected = false;
    static volatile boolean inventoryRunning = false;

    public static void main(String[] args) throws IOException {
        int port = args.length > 0 ? Integer.parseInt(args[0]) : 3010;
        HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", port), 0);
        server.setExecutor(Executors.newSingleThreadExecutor()); // serialize SDK access

        server.createContext("/connect", ex -> handle(ex, q -> {
            String ip = q.getOrDefault("ip", "192.168.254.202");
            int p = Integer.parseInt(q.getOrDefault("port", "8888"));
            boolean ok = uhf.init(ip, p);
            connected = ok;
            return "{\"ok\":" + ok + "}";
        }));

        server.createContext("/disconnect", ex -> handle(ex, q -> {
            if (inventoryRunning) { uhf.stopInventory(); inventoryRunning = false; }
            boolean ok = uhf.free();
            connected = false;
            return "{\"ok\":" + ok + "}";
        }));

        server.createContext("/status", ex -> handle(ex, q ->
            "{\"ok\":true,\"connected\":" + connected + ",\"inventory\":" + inventoryRunning +
            ",\"queued\":" + tagQueue.size() + "}"));

        server.createContext("/version", ex -> handle(ex, q -> {
            String v = uhf.getVersion();
            return "{\"ok\":" + (v != null) + ",\"version\":" + jstr(v) + "}";
        }));

        server.createContext("/inventory/start", ex -> handle(ex, q -> {
            // mode=buffer skips the callback and relies on readTagFromBuffer
            // (the SDK offers exactly one of the two receive paths per run).
            boolean useCallback = !"buffer".equals(q.get("mode"));
            if (useCallback) {
                uhf.setInventoryCallback(info -> {
                    String tag = tagJson(info);
                    tagQueue.add(tag);
                    while (tagQueue.size() > 5000) tagQueue.poll(); // bound memory
                });
            } else {
                uhf.setInventoryCallback(null);
            }
            boolean ok = uhf.startInventoryTag();
            inventoryRunning = ok;
            return "{\"ok\":" + ok + ",\"receive\":\"" + (useCallback ? "callback" : "buffer") + "\"}";
        }));

        // drain via the deprecated buffer API (works when started with mode=buffer)
        server.createContext("/tags-buffer", ex -> handle(ex, q -> {
            int max = Integer.parseInt(q.getOrDefault("max", "200"));
            StringBuilder sb = new StringBuilder("{\"ok\":true,\"tags\":[");
            int n = 0;
            UHFTAGInfo info;
            while (n < max && (info = uhf.readTagFromBuffer()) != null) {
                if (n++ > 0) sb.append(',');
                sb.append(tagJson(info));
            }
            return sb.append("]}").toString();
        }));

        server.createContext("/inventory/stop", ex -> handle(ex, q -> {
            boolean ok = uhf.stopInventory();
            inventoryRunning = false;
            return "{\"ok\":" + ok + "}";
        }));

        server.createContext("/tags", ex -> handle(ex, q -> {
            int max = Integer.parseInt(q.getOrDefault("max", "200"));
            StringBuilder sb = new StringBuilder("{\"ok\":true,\"tags\":[");
            String t;
            int n = 0;
            while (n < max && (t = tagQueue.poll()) != null) {
                if (n++ > 0) sb.append(',');
                sb.append(t);
            }
            return sb.append("]}").toString();
        }));

        server.createContext("/gpi", ex -> handle(ex, q -> {
            List<GPIStateEntity> list = uhf.getGPI();
            if (list == null) return "{\"ok\":false}";
            StringBuilder sb = new StringBuilder("{\"ok\":true,\"gpi\":[");
            for (int i = 0; i < list.size(); i++) {
                GPIStateEntity g = list.get(i);
                if (i > 0) sb.append(',');
                sb.append("{\"name\":").append(jstr(g.getGPIName()))
                  .append(",\"state\":").append(g.getGPIState()).append('}');
            }
            return sb.append("]}").toString();
        }));

        server.createContext("/power", ex -> handle(ex, q -> {
            if ("POST".equals(ex.getRequestMethod())) {
                int dbm = Integer.parseInt(q.get("dbm"));
                boolean all = true;
                for (String s : q.getOrDefault("ants", "1").split(",")) {
                    AntennaNameEnum ant = AntennaNameEnum.getValue(Integer.parseInt(s.trim()));
                    if (ant != null) all &= uhf.setPower(ant, dbm);
                }
                return "{\"ok\":" + all + "}";
            }
            List<AntennaPowerEntity> list = uhf.getPowerAll();
            if (list == null) return "{\"ok\":false}";
            StringBuilder sb = new StringBuilder("{\"ok\":true,\"power\":[");
            for (int i = 0; i < list.size(); i++) {
                AntennaPowerEntity p = list.get(i);
                if (i > 0) sb.append(',');
                sb.append("{\"ant\":").append(p.getAntennaNameEnum().getValue())
                  .append(",\"dbm\":").append(p.getPower()).append('}');
            }
            return sb.append("]}").toString();
        }));

        // Which ports have an antenna PHYSICALLY connected. See AntennaLink.
        //
        // Never asked while tags are flowing. An antenna only changes when
        // somebody unplugs one, so there is no reason for a monitoring question
        // to share the wire with a pallet going through — it answers `busy` and
        // the caller keeps its previous reading. This is why the check does NOT
        // use withInventoryPaused the way GET /antennas does: pausing a live
        // gate to light a status dot is the wrong trade.
        server.createContext("/antennas/link", ex -> handle(ex, q -> {
            if (!connected) return "{\"ok\":false,\"error\":\"not connected\"}";
            if (inventoryRunning) return "{\"ok\":false,\"busy\":true}";
            byte[] raw = com.rscja.deviceapi.AntennaLink.ask(uhf);
            if (raw == null) return "{\"ok\":false,\"error\":\"no reply from reader\"}";
            return "{\"ok\":true,\"raw\":" + jstr(hex(raw)) + "}";
        }));

        server.createContext("/antennas", ex -> handle(ex, q -> {
            if ("POST".equals(ex.getRequestMethod())) {
                List<AntennaState> states = new ArrayList<>();
                List<Integer> want = new ArrayList<>();
                for (String s : q.getOrDefault("ports", "1").split(",")) want.add(Integer.parseInt(s.trim()));
                for (int i = 1; i <= 16; i++) {
                    AntennaNameEnum ant = AntennaNameEnum.getValue(i);
                    if (ant != null) states.add(new AntennaState(ant, want.contains(i)));
                }
                return "{\"ok\":" + uhf.setAntenna(states) + "}";
            }
            List<AntennaState> list = uhf.getAntenna();
            if (list == null) return "{\"ok\":false}";
            StringBuilder sb = new StringBuilder("{\"ok\":true,\"enabled\":[");
            boolean first = true;
            for (AntennaState s : list) {
                if (s.isEnable()) {
                    if (!first) sb.append(',');
                    sb.append(s.getAntennaName().getValue());
                    first = false;
                }
            }
            return sb.append("]}").toString();
        }));

        server.createContext("/tag/single", ex -> handle(ex, q -> {
            UHFTAGInfo info = uhf.inventorySingleTag();
            if (info == null) return "{\"ok\":true,\"tag\":null}";
            return "{\"ok\":true,\"tag\":{\"pc\":" + jstr(info.getPc()) +
                   ",\"epc\":" + jstr(info.getEPC()) + "}}";
        }));

        server.createContext("/tag/read", ex -> handle(ex, q -> {
            String pwd = q.getOrDefault("pwd", "00000000");
            int bank = Integer.parseInt(q.get("bank"));
            int ptr = Integer.parseInt(q.getOrDefault("ptr", "0"));
            int words = Integer.parseInt(q.getOrDefault("words", "1"));
            String fdata = q.get("fdata");
            String hex;
            if (fdata == null || fdata.isEmpty()) {
                hex = uhf.readData(pwd, bank, ptr, words);
            } else {
                int fbank = Integer.parseInt(q.getOrDefault("fbank", "1"));
                int fptr = Integer.parseInt(q.getOrDefault("fptr", "32")); // bits
                hex = uhf.readData(pwd, fbank, fptr, fdata.length() * 4, fdata, bank, ptr, words);
            }
            return "{\"ok\":" + (hex != null) + ",\"hex\":" + jstr(hex) + "}";
        }));

        server.createContext("/tag/write", ex -> handle(ex, q -> {
            String pwd = q.getOrDefault("pwd", "00000000");
            int bank = Integer.parseInt(q.get("bank"));
            int ptr = Integer.parseInt(q.getOrDefault("ptr", "0"));
            String data = q.get("data");
            if (data == null || data.isEmpty() || data.length() % 4 != 0)
                return "{\"ok\":false,\"error\":\"data must be whole 16-bit words of hex\"}";
            int words = data.length() / 4;
            String fdata = q.get("fdata");
            boolean ok;
            if (fdata == null || fdata.isEmpty()) {
                ok = uhf.writeData(pwd, bank, ptr, words, data);
            } else {
                int fbank = Integer.parseInt(q.getOrDefault("fbank", "1"));
                int fptr = Integer.parseInt(q.getOrDefault("fptr", "32")); // bits
                ok = uhf.writeData(pwd, fbank, fptr, fdata.length() * 4, fdata, bank, ptr, words, data);
            }
            return "{\"ok\":" + ok + "}";
        }));

        server.createContext("/workmode", ex -> handle(ex, q -> {
            if ("POST".equals(ex.getRequestMethod())) {
                int mode = Integer.parseInt(q.get("mode"));
                return "{\"ok\":" + uhf.setWorkMode(mode) + "}";
            }
            return "{\"ok\":true,\"mode\":" + uhf.getWorkMode() + "}";
        }));

        // Reader buzzer. The UR4 chirps on every single tag read out of the box,
        // which at a gate reading continuously is a solid tone all shift — so
        // this exists purely to shut it up: POST /beep?on=0.
        //
        // Persists in the reader, so it survives a power cycle and does NOT need
        // re-applying on connect.
        server.createContext("/beep", ex -> handle(ex, q -> {
            if ("POST".equals(ex.getRequestMethod())) {
                int on = Integer.parseInt(q.getOrDefault("on", "0"));
                return "{\"ok\":" + uhf.setBeep(on) + "}";
            }
            // getBeep returns the raw status characters; first one carries the
            // on/off flag. Null means the reader would not answer, which is not
            // the same as "off" and must not be reported as it.
            char[] b = uhf.getBeep();
            if (b == null || b.length == 0) return "{\"ok\":false}";
            return "{\"ok\":true,\"on\":" + ((int) b[0] != 0) + ",\"raw\":" + (int) b[0] + "}";
        }));

        server.start();
        System.out.println("[sidecar] listening on http://127.0.0.1:" + port);
    }

    // --- plumbing ---------------------------------------------------------

    interface Handler { String run(Map<String, String> query) throws Exception; }

    static void handle(HttpExchange ex, Handler h) throws IOException {
        String body;
        int code = 200;
        try {
            body = h.run(parseQuery(ex.getRequestURI().getRawQuery()));
        } catch (Exception e) {
            code = 500;
            body = "{\"ok\":false,\"error\":" + jstr(String.valueOf(e)) + "}";
        }
        byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
        ex.getResponseHeaders().set("Content-Type", "application/json");
        ex.sendResponseHeaders(code, bytes.length);
        try (OutputStream os = ex.getResponseBody()) { os.write(bytes); }
    }

    static Map<String, String> parseQuery(String raw) {
        Map<String, String> m = new HashMap<>();
        if (raw == null) return m;
        for (String kv : raw.split("&")) {
            int i = kv.indexOf('=');
            if (i > 0) m.put(kv.substring(0, i), java.net.URLDecoder.decode(kv.substring(i + 1), StandardCharsets.UTF_8));
        }
        return m;
    }

    static String tagJson(UHFTAGInfo info) {
        return "{\"epc\":" + jstr(info.getEPC()) +
               ",\"pc\":" + jstr(info.getPc()) +
               ",\"tid\":" + jstr(info.getTid()) +
               ",\"user\":" + jstr(info.getUser()) +
               ",\"ant\":" + jstr(info.getAnt()) +
               ",\"rssi\":" + jstr(info.getRssi()) +
               ",\"ts\":" + System.currentTimeMillis() + "}";
    }

    /** Bytes as uppercase hex, so the Node side can parse a reply this build has never seen. */
    static String hex(byte[] b) {
        StringBuilder sb = new StringBuilder(b.length * 2);
        for (byte x : b) sb.append(String.format("%02X", x));
        return sb.toString();
    }

    static String jstr(String s) {
        if (s == null) return "null";
        StringBuilder sb = new StringBuilder("\"");
        for (char c : s.toCharArray()) {
            switch (c) {
                case '"': sb.append("\\\""); break;
                case '\\': sb.append("\\\\"); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                case '\t': sb.append("\\t"); break;
                default:
                    if (c < 0x20) sb.append(String.format("\\u%04x", (int) c));
                    else sb.append(c);
            }
        }
        return sb.append('"').toString();
    }
}
